stacks-signer/src/runloop.rs

```diff
@@ -300,10 +345,177 @@ impl<C: Coordinator> RunLoop<C> {
+    /// If the request is for a block it has already agreed to sign, it will overwrite the message with the agreed upon value
+    /// Returns whether the request is valid or not.
+    fn validate_signature_share_request(&self, request: &mut SignatureShareRequest) -> bool {
+        // A coordinator could have sent a signature share request with a different message than we agreed to sign
+        match self
+            .blocks
+            .get(&request.message)
+            .map(|block_info| &block_info.vote)
+        {
+            Some(Some(vote)) => {
+                // Overwrite with our agreed upon value in case another message won majority or the coordinator is trying to cheat...
+                request.message = vote.clone();
+                true
+            }
+            Some(None) => {
+                // We have seen this block before, but we have not agreed to sign it. TODO: ignore it
+                debug!("Received a signature share request for a block we have not validated yet.");
+                false
+            }
+            None => {
+                // We have not seen this block before.
+                // TODO: should probably ignore any messages that are not either sBTC transactons or Nakamoto blocks. Leave now for abitrary message signing
+                debug!("Received a signature share request for an unknown message stream. Signing it as is...");
+                true
+            }
+        }
+    }
+
+    /// Helper function to validate a nonce request, updating its message appropriately.
+    /// Note that if the request is for a block, we will update the request message
+    /// as either a hash indicating a vote no or the signature hash indicating a vote yes
+    /// Returns whether the request is valid or not
+    fn validate_nonce_request(&mut self, request: &mut NonceRequest) -> bool {
+        let mut ptr = &request.message[..];
+        let Some(block) = read_next::<NakamotoBlock, _>(&mut ptr).ok() else {
+            // TODO: we should probably reject requests to sign things that are not blocks or transactions (leave for now to enable testing abitrary signing)
+            warn!("Received a nonce request for an unknown message stream. Signing the nonce request as is.");
+            return true;
+        };
+        let Ok(hash) = block.header.signature_hash() else {
+            debug!("Received a nonce request for a block with an invalid signature hash. Ignore it.");
+            return false;
+        };
+        let mut hash_bytes = hash.0.to_vec();
+        let transactions = &self.transactions;
+        let block_info = self.blocks.entry(hash_bytes.clone()).or_insert(BlockInfo {
+            vote: None,
+            valid: false,
+            block: block.clone(),
+        });
+        // Validate the block contents
+        block_info.valid = Self::validate_block(block_info, transactions);
```

Comment:
Don't we actually want to validate the block via the RPC endpoint here?  Or, is this already done by the time this code is reached?

Reply:
Assuming this is reached by `verify_chunk()`, I'm assuming that you still need to do the RPC call to the node?

Reply:
The current flow is this:

miner proposes block, all signers see it and immediately do a call to the rpc endpoint.

Signers receive their node endpoiint rpc responses as events and they update their seen block with this info.

Coordinator gets the node endpoint rpc response and triggers a signing round

Signers get the block sign request from coordinators and check against their map of blocks and see if they have validated it against the endpoint or not (will be false if they have not) -> signers will approve or deny accordingly

Signers and coordinators all broadcast the result of the signing round back to the miners.

Reply:
So I have updated the code to be a bit more resilient to out of order messages.

It will currently cache nonce requests on the side for blocks it has not yet validated. Responding to these nonce requests once it has the validated info returned. It is slightly janky how I currently store blocks on the side, but am hoping with @kantai 's ticket to properly handle state, this will go away